src/main/java/edu/texas/social_computing/hospitals/RHRP.java

```diff
@@ -0,0 +1,134 @@
+package edu.texas.social_computing.hospitals;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.io.FileNotFoundException;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+public class RHRP {
+    private RHRP() {
+    }
+
+    /**
+     * Real HRP with Couples From http://www.nrmp.org/couples-match-videos/
+     * - Treat couples as a unit
+     * - Only make a match if the hospital is on resident's pref list and resident is on hospital's
+     * - If one partner has failed to match, increment the other's pref rank
+     * - If one partner has been kicked out due to being the worst, then break the other partner's match (it it has one)
+     * - Assumes that both partners have same pref list size
+     */
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents,
+                               Matching existingMatching) {
+        Matching m = existingMatching;
+
+        while (!freeResidents.isEmpty()) {
+            Resident currentResident = freeResidents.poll();
+            if (currentResident.getPreferences().isEmpty()) {
```

Comment:
it probably makes sense to move this sort of thing to a validateInput method that gets called at the start of `RHRP.run()` or something. this took me a while to understand what was going on here.
```
    private boolean hasValidCouples(ResidentTable residentTable) {
        for (Resident resident : residentTable.getAll()) {
            if (resident.hasPartner()) {
                Resident partner = residentTable.getResidentById(resident.getPartnerId());
                if (resident.getInitialPreferences().size() != partner.getInitialPreferences().size()) {
                    return false;
                }
            }
        }
        return true;
    }
```

Reply:
I like this idea.

Reply:
Yes, before runnig this alg it is probably better to check if the inputs are valid, but I was doing on a case-by-case basis. Your method makes it cleaner.